sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SetFns.java

```diff
@@ -187,83 +237,112 @@
-        ParDo.of(
-            new DoFn<KV<T, CoGbkResult>, T>() {
-
-              @ProcessElement
-              public void processElement(ProcessContext c) {
-                KV<T, CoGbkResult> elementGroups = c.element();
-
-                CoGbkResult value = elementGroups.getValue();
-                long inFirstSize = Iterables.size(value.getAll(leftCollectionTag));
-                long inSecondSize = Iterables.size(value.getAll(rightCollectionTag));
-
-                T element = elementGroups.getKey();
-                for (long i = 0L; i < fn.apply(inFirstSize, inSecondSize); i++) {
-                  c.output(element);
-                }
-              }
-            }));
+  public static <T> Flatten.PCollections<T> unionAll() {
+    return Flatten.pCollections();
   }
 
   public static class SetImpl<T> extends PTransform<PCollection<T>, PCollection<T>> {
-    private final PCollection<T> rightCollection;
-    private final SerializableBiFunction<Long, Long, Long> fn;
 
-    private SetImpl(PCollection<T> rightCollection, SerializableBiFunction<Long, Long, Long> fn) {
-      this.rightCollection = rightCollection;
-      this.fn = fn;
+    private final transient PCollection<T> right;
+    private final PTransform<PCollectionList<T>, PCollection<T>> listTransformFn;
+
+    private SetImpl(
+        PCollection<T> rightCollection,
+        PTransform<PCollectionList<T>, PCollection<T>> listTransformFn) {
+      this.right = rightCollection;
+      this.listTransformFn = listTransformFn;
     }
 
     @Override
     public PCollection<T> expand(PCollection<T> leftCollection) {
-      return performSetOperation(leftCollection, rightCollection, fn)
-          .setCoder(leftCollection.getCoder());
+      return PCollectionList.of(leftCollection).and(right).apply(listTransformFn);
     }
   }
 
-  public static class SetUnionAllImpl<T> extends PTransform<PCollection<T>, PCollection<T>> {
-    private final PCollection<T> rightCollection;
+  public static class SetImplCollections<T> extends PTransform<PCollectionList<T>, PCollection<T>> {
+
+    private final transient SerializableBiFunction<Long, Long, Long> fn;
+
+    private SetImplCollections(SerializableBiFunction<Long, Long, Long> fn) {
+      this.fn = fn;
+    }
+
+    private static <T> PCollection<T> performSetOperationCollectionList(
+            PCollectionList<T> inputs, SerializableBiFunction<Long, Long, Long> fn) {
+      List<PCollection<T>> all = inputs.getAll();
+      int size = all.size();
+      if (size == 1) {
+        return inputs.get(0); // Handle only one PCollection in list. Coder is already specified
+      } else {
+        PCollection<T> accum = inputs.get(0);
+        List<PCollection<T>> tail = all.subList(1, size);
 
-    private SetUnionAllImpl(PCollection<T> rightCollection) {
-      this.rightCollection = rightCollection;
+        for (PCollection<T> second : tail) {
+          accum = performSetOperation(accum, second, fn);
```

Comment:
It'd be better to implement the binary operations in terms of n-ary operations, so that one only need to so a single [Co]GBK. The operations are the same (e.g. union -> sum or 1, intersection = min or max(min(...), 1) for all/distinct variants respectively.

Reply:
Thanks. Thats very good idea. I have done it.

Reply:
renamed all methods with explicit Distinct and All as prefix